plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -517,6 +517,30 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
          });
       }
 
+      // Returns contract name, action name, and exception text
+      // of an exception occured in a contract
+      std::tuple<account_name, action_name, fc::string> get_detailed_contract_exception_info(const fc::exception_ptr& except_ptr,
+                                     const std::variant<fc::exception_ptr, transaction_trace_ptr>& response) {
+         account_name contract_name;
+         action_name  act_name;
+         fc::string   details;
+
+         if (std::holds_alternative<transaction_trace_ptr>(response)) {
+            auto resp_ptr = std::get<transaction_trace_ptr>(response);
+            if (resp_ptr && !resp_ptr->action_traces.empty()) {
+               fc::unsigned_int last_action_ordinal = resp_ptr->action_traces.size() - 1;
+               contract_name = resp_ptr->action_traces[last_action_ordinal].act.account;
+               act_name = resp_ptr->action_traces[last_action_ordinal].act.name;
+            }
+         }
+
+         if (except_ptr) {
+            details = fc::format_string("${d}", fc::mutable_variant_object() ("d", except_ptr->top_message()), true);  // true for limitting the formatted string size
```

Comment:
This seems like it is an inefficient way to call `variant.cpp` `clean_append`. Unfortunately, `clean_append` is only in `variant.cpp`. I think we should create a general purpose `clean` method that optionally escapes control characters (specifically newline) and truncates. That could then be used here. 
As is this leaves any newlines put in by the contract writer in the output.

Reply:
Can we do this later after the decision about FC de-submodule is made? I will create a Jira issue about it.